File: packages/runtime-web-three/src/navigation.ts
import type { IWorldIr, Vec3 } from "@threenative/ir";

export interface INavigationPathRequest {
  goal: Vec3;
  id?: string;
  start: Vec3;
}

export interface INavigationPathResult {
  failureReason?: "goal-outside" | "no-route" | "start-outside";
  path: Vec3[];
  query: string;
  status: "failed" | "success";
  totalCost: number;
  visitedRegions: string[];
}

interface INavigationRegion {
  area?: string;
  center: Vec3;
  id: string;
  neighbors?: string[];
  points: Array<readonly [number, number]>;
}

interface INavigationResource {
  areaCosts?: Record<string, number>;
  queries?: INavigationPathRequest[];
  regions: INavigationRegion[];
}

export function traceNavigationPaths(world: IWorldIr, requests?: ReadonlyArray<INavigationPathRequest>): INavigationPathResult[] {
  const navigation = world.resources?.Navigation as INavigationResource | undefined;
  if (navigation === undefined) {
    return [];
  }
  const queries = requests ?? navigation.queries ?? [];
  return queries.map((query, index) => pathQuery(navigation, query, query.id ?? `query-${index}`));
}

export function queryNavigationPath(world: IWorldIr, request: INavigationPathRequest): INavigationPathResult {
  const navigation = world.resources?.Navigation as INavigationResource | undefined;
  if (navigation === undefined) {
    return { failureReason: "no-route", path: [], query: request.id ?? "query", status: "failed", totalCost: 0, visitedRegions: [] };
  }
  return pathQuery(navigation, request, request.id ?? "query");
}

function pathQuery(navigation: INavigationResource, request: INavigationPathRequest, queryId: string): INavigationPathResult {
  const startRegion = regionForPoint(navigation.regions, request.start);
  if (startRegion === undefined) {
    return { failureReason: "start-outside", path: [], query: queryId, status: "failed", totalCost: 0, visitedRegions: [] };
  }
  const goalRegion = regionForPoint(navigation.regions, request.goal);
  if (goalRegion === undefined) {
    return { failureReason: "goal-outside", path: [], query: queryId, status: "failed", totalCost: 0, visitedRegions: [startRegion.id] };
  }
  const route = shortestRoute(navigation, startRegion.id, goalRegion.id);
  if (route.length === 0) {
    return { failureReason: "no-route", path: [], query: queryId, status: "failed", totalCost: 0, visitedRegions: [startRegion.id] };
  }
  return {
      path: [roundVec3(request.start), ...route.slice(1, -1).map((id) => roundVec3(regionById(navigation.regions, id)!.center)), roundVec3(request.goal)],
    query: queryId,
    status: "success",
    totalCost: round(routeCost(navigation, route)),
    visitedRegions: route,
  };
}

function shortestRoute(navigation: INavigationResource, start: string, goal: string): string[] {
  const regions = new Map(navigation.regions.map((region) => [region.id, region]));
  const costs = new Map<string, number>([[start, 0]]);
  const previous = new Map<string, string>();
  const queue = [...regions.keys()].sort((left, right) => left.localeCompare(right));
  while (queue.length > 0) {
    queue.sort((left, right) => (costs.get(left) ?? Number.POSITIVE_INFINITY) - (costs.get(right) ?? Number.POSITIVE_INFINITY) || left.localeCompare(right));
    const current = queue.shift()!;
    if (current === goal) {
      break;
    }
    const currentCost = costs.get(current);
    if (currentCost === undefined) {
      continue;
    }
    for (const neighbor of [...(regions.get(current)?.neighbors ?? [])].sort((left, right) => left.localeCompare(right))) {
      if (!regions.has(neighbor)) {
        continue;
      }
      const candidate = currentCost + regionCost(navigation, regions.get(neighbor)!);
      if (candidate < (costs.get(neighbor) ?? Number.POSITIVE_INFINITY)) {
        costs.set(neighbor, candidate);
        previous.set(neighbor, current);
      }
    }
  }
  if (!costs.has(goal)) {
    return [];
  }
  const route = [goal];
  while (route[0] !== start) {
    const prior = previous.get(route[0]!);
    if (prior === undefined) {
      return [];
    }
    route.unshift(prior);
  }
  return route;
}

function routeCost(navigation: INavigationResource, route: readonly string[]): number {
  return route.slice(1).reduce((total, id) => {
    const region = regionById(navigation.regions, id);
    return region === undefined ? total : total + regionCost(navigation, region);
  }, 0);
}

function regionCost(navigation: INavigationResource, region: INavigationRegion): number {
  return navigation.areaCosts?.[region.area ?? "default"] ?? 1;
}

function regionForPoint(regions: readonly INavigationRegion[], point: Vec3): INavigationRegion | undefined {
  return [...regions].sort((left, right) => left.id.localeCompare(right.id)).find((region) => pointInPolygon([point[0], point[2]], region.points));
}

function regionById(regions: readonly INavigationRegion[], id: string): INavigationRegion | undefined {
  return regions.find((region) => region.id === id);
}

function pointInPolygon(point: readonly [number, number], polygon: ReadonlyArray<readonly [number, number]>): boolean {
  let inside = false;
  for (let index = 0, previous = polygon.length - 1; index < polygon.length; previous = index++) {
    const currentPoint = polygon[index];
    const previousPoint = polygon[previous];
    if (currentPoint === undefined || previousPoint === undefined) {
      continue;
    }
    const intersects = ((currentPoint[1] > point[1]) !== (previousPoint[1] > point[1]))
      && point[0] < (previousPoint[0] - currentPoint[0]) * (point[1] - currentPoint[1]) / (previousPoint[1] - currentPoint[1]) + currentPoint[0];
    if (intersects) {
      inside = !inside;
    }
  }
  return inside;
}

function roundVec3(value: Vec3): Vec3 {
  return [round(value[0]), round(value[1]), round(value[2])];
}

function round(value: number): number {
  return Math.round(value * 1_000_000) / 1_000_000;
}

File: runtime-bevy/crates/threenative_runtime/src/navigation.rs
use serde::{Deserialize, Serialize};
use threenative_loader::LoadedBundle;

#[derive(Clone, Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct NavigationPathRequest {
    pub goal: [f32; 3],
    pub id: Option<String>,
    pub start: [f32; 3],
}

#[derive(Clone, Debug, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct NavigationPathResult {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub failure_reason: Option<String>,
    pub path: Vec<[f32; 3]>,
    pub query: String,
    pub status: String,
    pub total_cost: f32,
    pub visited_regions: Vec<String>,
}

#[derive(Clone, Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
struct NavigationResource {
    #[serde(default)]
    area_costs: std::collections::BTreeMap<String, f32>,
    #[serde(default)]
    queries: Vec<NavigationPathRequest>,
    regions: Vec<NavigationRegion>,
}

#[derive(Clone, Debug, Deserialize)]
struct NavigationRegion {
    area: Option<String>,
    center: [f32; 3],
    id: String,
    #[serde(default)]
    neighbors: Vec<String>,
    points: Vec<[f32; 2]>,
}

pub fn trace_navigation_paths(bundle: &LoadedBundle) -> Vec<NavigationPathResult> {
    let Some(navigation) = navigation_resource(bundle) else {
        return Vec::new();
    };
    navigation
        .queries
        .iter()
        .enumerate()
        .map(|(index, request)| path_query(&navigation, request, request.id.clone().unwrap_or_else(|| format!("query-{index}"))))
        .collect()
}

pub fn query_navigation_path(bundle: &LoadedBundle, request: &NavigationPathRequest) -> NavigationPathResult {
    let Some(navigation) = navigation_resource(bundle) else {
        return NavigationPathResult {
            failure_reason: Some("no-route".to_owned()),
            path: Vec::new(),
            query: request.id.clone().unwrap_or_else(|| "query".to_owned()),
            status: "failed".to_owned(),
            total_cost: 0.0,
            visited_regions: Vec::new(),
        };
    };
    path_query(&navigation, request, request.id.clone().unwrap_or_else(|| "query".to_owned()))
}

fn navigation_resource(bundle: &LoadedBundle) -> Option<NavigationResource> {
    serde_json::from_value(bundle.world.resources.get("Navigation")?.clone()).ok()
}

fn path_query(navigation: &NavigationResource, request: &NavigationPathRequest, query: String) -> NavigationPathResult {
    let start = region_for_point(&navigation.regions, request.start);
    let Some(start_region) = start else {
        return failed(query, "start-outside", Vec::new());
    };
    let goal = region_for_point(&navigation.regions, request.goal);
    let Some(goal_region) = goal else {
        return failed(query, "goal-outside", vec![start_region.id.clone()]);
    };
    let route = shortest_route(navigation, &start_region.id, &goal_region.id);
    if route.is_empty() {
        return failed(query, "no-route", vec![start_region.id.clone()]);
    }
    let mut path = vec![round_vec3(request.start)];
    for id in route.iter().skip(1).take(route.len().saturating_sub(2)) {
        if let Some(region) = navigation.regions.iter().find(|region| &region.id == id) {
            path.push(round_vec3(region.center));
        }
    }
    path.push(round_vec3(request.goal));
    NavigationPathResult {
        failure_reason: None,
        path,
        query,
        status: "success".to_owned(),
        total_cost: round(route_cost(navigation, &route)),
        visited_regions: route,
    }
}

fn failed(query: String, reason: &str, visited_regions: Vec<String>) -> NavigationPathResult {
    NavigationPathResult {
        failure_reason: Some(reason.to_owned()),
        path: Vec::new(),
        query,
        status: "failed".to_owned(),
        total_cost: 0.0,
        visited_regions,
    }
}

fn shortest_route(navigation: &NavigationResource, start: &str, goal: &str) -> Vec<String> {
    let regions = navigation.regions.iter().map(|region| (region.id.clone(), region)).collect::<std::collections::BTreeMap<_, _>>();
    let mut costs = std::collections::BTreeMap::from([(start.to_owned(), 0.0)]);
    let mut previous = std::collections::BTreeMap::<String, String>::new();
    let mut queue = regions.keys().cloned().collect::<Vec<_>>();
    while !queue.is_empty() {
        queue.sort_by(|left, right| {
            costs
                .get(left)
                .unwrap_or(&f32::INFINITY)
                .partial_cmp(costs.get(right).unwrap_or(&f32::INFINITY))
                .unwrap_or(std::cmp::Ordering::Equal)
                .then(left.cmp(right))
        });
        let current = queue.remove(0);
        if current == goal {
            break;
        }
        let Some(current_cost) = costs.get(&current).copied() else {
            continue;
        };
        let Some(region) = regions.get(&current) else {
            continue;
        };
        for neighbor in &region.neighbors {
            let Some(neighbor_region) = regions.get(neighbor) else {
                continue;
            };
            let candidate = current_cost + region_cost(navigation, neighbor_region);
            if candidate < *costs.get(neighbor).unwrap_or(&f32::INFINITY) {
                costs.insert(neighbor.clone(), candidate);
                previous.insert(neighbor.clone(), current.clone());
            }
        }
    }
    if !costs.contains_key(goal) {
        return Vec::new();
    }
    let mut route = vec![goal.to_owned()];
    while route.first().is_some_and(|id| id != start) {
        let Some(prior) = previous.get(route.first().unwrap()).cloned() else {
            return Vec::new();
        };
        route.insert(0, prior);
    }
    route
}

fn route_cost(navigation: &NavigationResource, route: &[String]) -> f32 {
    route
        .iter()
        .skip(1)
        .filter_map(|id| navigation.regions.iter().find(|region| &region.id == id))
        .map(|region| region_cost(navigation, region))
        .sum()
}

fn region_cost(navigation: &NavigationResource, region: &NavigationRegion) -> f32 {
    navigation.area_costs.get(region.area.as_deref().unwrap_or("default")).copied().unwrap_or(1.0)
}

fn region_for_point(regions: &[NavigationRegion], point: [f32; 3]) -> Option<&NavigationRegion> {
    let mut ordered = regions.iter().collect::<Vec<_>>();
    ordered.sort_by(|left, right| left.id.cmp(&right.id));
    ordered.into_iter().find(|region| point_in_polygon([point[0], point[2]], &region.points))
}

fn point_in_polygon(point: [f32; 2], polygon: &[[f32; 2]]) -> bool {
    let mut inside = false;
    let mut previous = polygon.len() - 1;
    for index in 0..polygon.len() {
        let current = polygon[index];
        let prior = polygon[previous];
        let intersects = ((current[1] > point[1]) != (prior[1] > point[1]))
            && point[0] < (prior[0] - current[0]) * (point[1] - current[1]) / (prior[1] - current[1]) + current[0];
        if intersects {
            inside = !inside;
        }
        previous = index;
    }
    inside
}

fn round_vec3(value: [f32; 3]) -> [f32; 3] {
    [round(value[0]), round(value[1]), round(value[2])]
}

fn round(value: f32) -> f32 {
    (value * 1_000_000.0).round() / 1_000_000.0
}
